### ISSUE TITLE
MACRO min/max under WINDOWS

### ISSUE THREAD

**Issue by amigo421** [jarro2783/cxxopts#49]
```
C:\dev\workspace\z_eph\tool\utils\cxxopts.hpp(1342): error C2589: '(': illegal token on right side of '::' 
C:\dev\workspace\z_eph\tool\utils\cxxopts.hpp(1342): error C2062: type 'unknown-type' unexpected 
C:\dev\workspace\z_eph\tool\utils\cxxopts.hpp(1342): error C2059: syntax error: ')' 
C:\dev\workspace\z_eph\tool\utils\cxxopts.hpp(1346): error C2589: '(': illegal token on right side of '::' 
C:\dev\workspace\z_eph\tool\utils\cxxopts.hpp(1346): error C2062: type 'unknown-type' unexpected 
C:\dev\workspace\z_eph\tool\utils\cxxopts.hpp(1346): error C2059: syntax error: ')' 

```
the errors related with min and max in the library
cxxopts.hpp:

```
            longest = std::max(longest, stringLength(s));
            format.push_back(std::make_pair(s, String()));
        }

        longest = std::min(longest, static_cast<size_t>(OPTION_LONGEST));

```

the problem (Windows specifics) usually related with that there are macro definitions for min and max.
this (below) solves the problem : 

![image](https://cloud.githubusercontent.com/assets/3259933/26396175/18b3229a-407b-11e7-935e-437e901aaab5.png)

I'll perform some tests and if this woks okay and will raise a PR with WINDOWS only ifdef

**Comment by jarro2783** [comment]
Which version of the compiler? I've had a few other people discuss MSVC and they haven't had issues.

Does it also work if you an extra set of parentheses around the argument to min and max?

**Comment by amigo421** [comment]
extra parentheses doesn't solve the problem, 
VC toolchain is version 15, aka VisualStudio 2017
compiler version 19.XXXX

**Comment by jarro2783** [comment]
I'm happy to accept a pull request. I wonder if a better option would be to write `(std::max)(x, y)` as suggested by the StackOverflow answers, rather than `#undef`.

**Comment by amigo421** [comment]
sure, Ill raise a request

**Comment by supercairos** [comment]
I usually define `NOMINMAX` as preprocessor def in VS

**Comment by jarro2783** [comment]
Is this still an issue for you?

**Comment by supercairos** [comment]
I'm using cxxopts on Windows and have no issue to report.
The preprocessor flags on Windows does the trick.

**Comment by Robmaister** [comment]
I just picked this up for a project on Windows. Definitely use the built-in `NOMINMAX` define for `Windows.h` instead of trying to undef it manually.

If you undef, `#include <cxxopts.hpp>` would have side-effects for applications assuming that min and max are existing macros in later code, since the preprocessor has no scope.

Proper usage:

```
#ifndef NOMINMAX
#define NOMINMAX
#endif

#include <Windows.h>

#include <cxxopts.hpp>
```

An example that would break with `#undef`:

```
#include <Windows.h>
#include <cxxopts.hpp>

int main (int argc, char* argv[]) {
    return max(2, 3);
}
```

**Comment by amigo421** [comment]
agreed ,NOMINMAX prefered way and it works fine for me too
request can be closed without modification the library

**Comment by zhongjingjogy** [comment]
@jarro2783 I am sorry to bother the rest of others about a closed issue. As someone who tries to use `cxxopts` on windows with MSVC, I found it is very frustrated to deal with  `MACRO min/max under WINDOWS`. I should blame that MSVC introduce such an annoying feature that affects extensively. But I would sugguest that opensource library should provide more friendly solution to third party users. As an example, add parenthesis to anywhere you have `min` and `max` as it is suggested by [NIRO](https://stackoverflow.com/questions/11544073/how-do-i-deal-with-the-max-macro-in-windows-h-colliding-with-max-in-std)

```
(std::numeric_limits<size_type>::max)()
(std::min)
(std::max)
```

As it is said, 
> No need for the NOMINMAX macro in this case, plus you won't get compiler warnings

I wrote this in memory of my lost 30 minutes at this time.
